### PR TITLE
Redo of Design/lines #75

### PR DESCRIPTION
**Pull request by HaicoDorenbos**
Since in the PR of Design/lines #75 there were strange issues related to Travis, I decided to recreate this pull request. For more information look at: https://github.com/RoboTeamTwente/roboteam_utils/pull/75

### CODE REVIEW

**Comment by HaicoDorenbos** [comment]
Regarding @RobotJesse comment about the dead tag. I agree to use another tag, but it should satisfy the following requirements:
- It should be big and noticeable, since otherwise it will not be seen and will cause future software developers to waste their time.
- It should be placed at all locations with dead code, both in the header and the source file, by the same reason as above.
- It should explain why the code was not removed, but marked with a tag instead, and should tell that not everyone fully agreed with this decision. Such that future team can decide for themselves what they want to do with the dead code. This is especially important, because future teams might get confused and think there is a very important reason why this code was kept (which is not the case).
- It should tell that the tag needs to be removed (and be removed at other places as well) if the code is used.

File: include/roboteam_utils/HalfLine.h [hunk 1]
```diff
@@ -6,31 +6,40 @@
...
         /**
-         * Construct a new HalfLine object
-         * @param start The location where the HalfLine starts
-         * @param direction This vector indicates in which direction the HalfLine is infinite long
+         * Construct a new HalfLine object by using 2 different points on that HalfLine. Make sure that the given points are different.
+         * @param start The location where the HalfLine starts.
+         * @param goesThrough An arbitrary other vector through the HalfLine goes (make sure that it is different than the start point).
          */
-        constexpr HalfLine(const Vector2 &start, const Vector2 &direction) noexcept : start{start}, direction{direction} {};
+        explicit HalfLine(const Vector2 &start, const Vector2 &goesThrough);
```
Comment (Yuhanun) [inline include/roboteam_utils/HalfLine.h:25]:
```suggestion
        contexpr HalfLine(const Vector2 &start, const Vector2 &goesThrough);
```

Doesn't have to be marked expliti if it has more than 2 arguments and constexpr ftw.

**Review by Yuhanun — CHANGES_REQUESTED** [review]

**Comment by HaicoDorenbos** [comment]
Some points mentioned in Design/float comparison #78 still need to be fixed.

File: include/roboteam_utils/HalfLine.h (hunk 1 above)
Comment (HaicoDorenbos) [reply include/roboteam_utils/HalfLine.h:25]: Causes compile errors, so I won't change it...

File: test/LineIntersectionTest.cpp [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+ * test cases where there could be a possibility of division by zero. Furthermore for the non-parallel test cases we create 3 classes based on the relative position where they
+ * intersect (or do not intersect): t < 0, 0 <= t <= 1, 1 < t. We extend these test cases with boundary value analysis, but for t is (minus) infinity we use a very small and
+ * very large number (we expect intersections to fail for FLT_MAX and -FLT_MAX). Moreover for parallel test cases we create 3 classes based on the LineSegment interpretation:
+ * 'Equal': both LineSegment interpretation are equal, 'Subset': one of the LineSegments interpretations is a subset of the other, 'Overlap': both LineSegments interpretations
+ * have a shared part, but non of them is a subset of the other one, 'Distinct': both LineSegments do not intersect, but have a shared finite LineSegment where both ar subsets
+ * of, 'Different': both LineSegments do not intersect and do not have a shared infinite Line where both are subsets of. Furthermore create a test case for two LineSegments
+ * that are 'Equal' points and that are 'Different' points.
+ */
```
Comment (RobotJente) [inline test/LineIntersectionTest.cpp]: I like this a lot!

File: test/LineIntersectionTest.cpp [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+ * vertical line (does not change in x direction), horizontal line (does not change in y direction), arbitrary other line (does change in both x and y direction). This is to
+ * test cases where there could be a possibility of division by zero. Furthermore for the non-parallel test cases we create 3 classes based on the relative position where they
+ * intersect (or do not intersect): t < 0, 0 <= t <= 1, 1 < t. We extend these test cases with boundary value analysis, but for t is (minus) infinity we use a very small and
+ * very large number (we expect intersections to fail for FLT_MAX and -FLT_MAX). Moreover for parallel test cases we create 3 classes based on the LineSegment interpretation:
+ * 'Equal': both LineSegment interpretation are equal, 'Subset': one of the LineSegments interpretations is a subset of the other, 'Overlap': both LineSegments interpretations
+ * have a shared part, but non of them is a subset of the other one, 'Distinct': both LineSegments do not intersect, but have a shared finite LineSegment where both ar subsets
```
Comment (RobotJente) [inline test/LineIntersectionTest.cpp]:
```suggestion
 * have a shared part, but non of them is a subset of the other one, 'Distinct': both LineSegments do not intersect, but have a shared finite LineSegment where both are subsets
```

File: test/PolygonTest.cpp [hunk 4]
```diff
@@ -96,9 +100,9 @@ TEST(isConvex, PolygonTest) {
                                         P1Last, P1First, P2First, P2Last, P3First, P3Last, P4};
     //P1Last,P1First,;
     // convex calculation only makes sense if the examples we use are actually valid
-    for (const Polygon &example : allExamples) {
-        ASSERT_TRUE(example.isSimple());
-    }
+    // for (const Polygon &example : allExamples) {
+    //     ASSERT_TRUE(example.isSimple());
+    // }
```
Comment (RobotJente) [inline test/PolygonTest.cpp]: I know this is WIP but I just wanted to point this out

**Review by RobotJente — COMMENTED** [review]
This PR is looking quite nice :) I see you removed some dead code in the Line class etc, is the functionality of that code still preserved?

File: test/PolygonTest.cpp (hunk 4 above)
Comment (HaicoDorenbos) [reply test/PolygonTest.cpp]: Fixed

**Comment by HaicoDorenbos** [comment]
> This PR is looking quite nice :) I see you removed some dead code in the Line class etc, is the functionality of that code still preserved?

No, because a lot in these classes has been completely redesigned. So I decided that the most easiest solution was to remove the dead code in the Line & LineSegment classes rather than remove bugs out of these dead methods, redesign and re-document these dead methods.

**Comment by HaicoDorenbos** [comment]
Everything should be finished...

File: include/roboteam_utils/HalfLine.h (hunk 1 above)
Comment (timovdk) [reply include/roboteam_utils/HalfLine.h:25]: You should remove explicit though, unless you have a reason for it

File: include/roboteam_utils/Line.h [hunk 5]
```diff
@@ -1,207 +1,93 @@
...
     /**
-     * @brief Gets a pair of coefficients
+     * Creates a new Line instance by using 2 different points on that Line. Make sure that the given points are different.
      *
-     * @return std::pair<double, double> Pair of doubles where .first == slope() and .second == intercept()
+     * @param v1 An arbitrary point on the Line.
+     * @param v2 An arbitrary different point on the Line (make sure that it is different than the v2 point).
      */
-    [[nodiscard]] std::pair<double, double> coefficients() const;
+    explicit Line(const Vector2 &v1, const Vector2 &v2);
```
Comment (timovdk) [inline include/roboteam_utils/Line.h]: Explicit is not necessary here
```suggestion
    Line(const Vector2 &v1, const Vector2 &v2);
```

File: include/roboteam_utils/LineSegment.h [hunk 6]
```diff
@@ -1,249 +1,132 @@
...
+         *
+         * @param line The other (finite) LineSegment.
+         * @return Returns a list of the intersections.
+         */
+        [[nodiscard]] std::vector<Vector2> multiIntersect(const LineSegment &line) const;
+
+        /**
+         * Destructor of the LineSegment.
+         */
+        virtual ~LineSegment() = default;
```
Comment (timovdk) [inline include/roboteam_utils/LineSegment.h:112]: Why is there a virtual destructor here? Am I missing something?

File: src/Line.cpp [hunk 7]
```diff
@@ -1,127 +1,70 @@
...
-bool Line::isParallel(const Line &line) const {
-    // check if line is vertical, otherwise check the slope
-    if (this->isVertical() || line.isVertical()) {
-        return this->isVertical() && line.isVertical();
+Line::Line(const Vector2 &v1, const Vector2 &v2) {
+    this->v1 = v1;
+    this->v2 = v2;
+    if (v1 == v2) {
+        std::cerr << "Warning: you created an undefined line, because v1 == v2. Note that Lines have an infinite length. If you want to have a Line with finite length then use "
+                     "the LineSegment class instead." << std::endl;
```
Comment (timovdk) [inline src/Line.cpp]: I have my doubts with how you are handling this. At least use the print utils Lukas made.

File: src/Line.cpp [hunk 8]
```diff
@@ -1,127 +1,70 @@
...
-    // check if line is vertical, otherwise check the slope
-    if (this->isVertical() || line.isVertical()) {
-        return this->isVertical() && line.isVertical();
+
+Line::Line(const LineSegment &other) noexcept {
+    v1 = other.start;
+    v2 = other.end;
+    if (v1 == v2) {
+        std::cerr << "Warning: you created an undefined line, because v1 == v2. Note that Lines have an infinite length. If you want to have a Line with finite length then use "
+                     "the LineSegment class instead." << std::endl;
```
Comment (timovdk) [inline src/Line.cpp]: Same here, print utils

File: src/LineSegment.cpp [hunk 9]
```diff
@@ -1,339 +1,135 @@
...
-        return false;
-    }
-    //check if the point is in between the two points
-    double dot = A.dot(B);
-    if (dot < 0) {
-        return false;
-    }
-    if (dot > A.dot(A)) {
-        return false;
+    if (Line(*this).isOnLine(point)) {
```
Comment (timovdk) [inline src/LineSegment.cpp:70]: So you are contructing this line inside of this line? or what

**Review by timovdk — CHANGES_REQUESTED** [review]
Some comments

File: include/roboteam_utils/LineSegment.h (hunk 6 above)
Comment (HaicoDorenbos) [reply include/roboteam_utils/LineSegment.h:112]: I did not add or remove the destructor. I only changed the ordening of the functions.

File: src/Line.cpp (hunk 7 above)
Comment (HaicoDorenbos) [reply src/Line.cpp]: If you show an example how to use it then I will do...

File: src/LineSegment.cpp (hunk 9 above)
Comment (HaicoDorenbos) [reply src/LineSegment.cpp:70]: No, you create a Line inside a LineSegment class. The idea of this is to re-use the functionality of the Line class inside the LineSegment class and prevent code clones...

File: include/roboteam_utils/HalfLine.h (hunk 1 above)
Comment (HaicoDorenbos) [reply include/roboteam_utils/HalfLine.h:25]: Fixed

**Comment by HaicoDorenbos** [comment]
Everything should be answered...

File: src/Line.cpp (hunk 7 above)
Comment (timovdk) [reply src/Line.cpp]: #include"Print.h"

Print("Warning: you created....")

**Comment by Yuhanun** [comment]
What's the compiler error caused by the constexpr flag on the ctor?

**Comment by Multifacio** [comment]
The constructor is defined in the source file. You cannot use a constexpr flag if you define something.

**Comment by HaicoDorenbos** [comment]
The constructor is defined in the source file, so it is not possible to use a constexpr flag.

**Comment by HaicoDorenbos** [comment]
The print warning are fixed by now

File: src/Tube.cpp [hunk 10]
```diff
@@ -8,17 +8,14 @@ namespace rtt{
...
-}
 bool Tube::isCircle() const {
     return lineSegment.isPoint();
 }
 Tube::Tube(const LineSegment& line, double radius) : lineSegment{line}, radius{radius} {
 }
 Tube::Tube(const Vector2& start, const Vector2& end, double radius) : lineSegment{LineSegment(start,end)},radius{radius} {
 }
-Tube::Tube() : lineSegment{},radius{1.0}{ // lines default construct to (0,0) to (0,0)
+Tube::Tube() : lineSegment{LineSegment({0, 0}, {0, 0})},radius{1.0}{ // lines default construct to (0,0) to (0,0)
```
Comment (Yuhanun) [inline src/Tube.cpp]:
```suggestion
Tube::Tube() : lineSegment{{0, 0}, {0, 0}},radius{1.0}{ // lines default construct to (0,0) to (0,0)
```

File: src/Polygon.cpp [hunk 11]
```diff
@@ -201,4 +178,26 @@ namespace rtt {
...
+        std::vector<LineSegment> lines{ };
+        size_t numberIntersections = 0;
+        for (auto first = vertices.begin(); first != vertices.end(); first++) {
+            std::optional<LineSegment> boundarySegment;
+            if (first == std::prev(vertices.end())) {
+                boundarySegment = LineSegment(*first, vertices[0]);
+            } else {
+                boundarySegment = LineSegment(*first, *(first + 1));
+            }
+            for (auto line : lines) {
```
Comment (Yuhanun) [inline src/Polygon.cpp]:
```suggestion
            for (auto const& line : lines) {
```

File: src/Polygon.cpp [hunk 12]
```diff
@@ -201,4 +178,26 @@ namespace rtt {
     Vector2 Polygon::verticeCentroid() const {
         return std::accumulate(vertices.begin(), vertices.end(), Vector2(0, 0)) /= vertices.size();
     }
+
+    bool Polygon::isSimple() const {
+        // we loop over every unique pair
+        std::vector<LineSegment> lines{ };
+        size_t numberIntersections = 0;
+        for (auto first = vertices.begin(); first != vertices.end(); first++) {
```
Comment (Yuhanun) [inline src/Polygon.cpp:186]: Do you realize this is a range-based for loop?

File: src/Line.cpp [hunk 13]
```diff
@@ -1,127 +1,71 @@
...
-}
-
-bool Line::isParallel(const Line &line) const {
-    // check if line is vertical, otherwise check the slope
-    if (this->isVertical() || line.isVertical()) {
-        return this->isVertical() && line.isVertical();
+Line::Line(const Vector2 &v1, const Vector2 &v2) {
+    this->v1 = v1;
+    this->v2 = v2;
+    if (v1 == v2) {
```
Comment (Yuhanun) [inline src/Line.cpp:9]: This is runtime overhead i don't want in release mode.

File: src/Line.cpp [hunk 14]
```diff
@@ -1,127 +1,71 @@
...
 }
-bool Line::isParallel(const LineSegment &line) const {
-    // check if line is vertical, otherwise check the slope
-    if (this->isVertical() || line.isVertical()) {
-        return this->isVertical() && line.isVertical();
+
+Line::Line(const LineSegment &other) noexcept {
+    v1 = other.start;
+    v2 = other.end;
+    if (v1 == v2) {
```
Comment (Yuhanun) [inline src/Line.cpp:18]: More runtime overhead.

**Review by Yuhanun — CHANGES_REQUESTED** [review]

File: src/Line.cpp (hunk 13 above)
Comment (HaicoDorenbos) [reply src/Line.cpp:9]: We should check for exceptions

File: src/Line.cpp (hunk 14 above)
Comment (HaicoDorenbos) [reply src/Line.cpp:18]: Exception checking

**Review by Yuhanun — APPROVED** [review]
LGTM
